src/tools/isabelle_parse_utils.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
+
+# TODO: this parser doesn't work well because it relies on newlines as delimiters, 
```

Comment:
This is the main point which can still be improved -- right now I use newlines as a delimiter, but a lot of Isabelle (Isar) proofs have tactics that are split up in weird ways over multiple lines (for readability).

This doesn't cause any problems for the agent since they don't look at the proof, but it may cause some issues later on. I think I'd like to see how the executor gets used elsewhere before deciding on the best fix. Just something to keep in mind